### PR TITLE
Update jenkins task to use item-level-downloader

### PR DESCRIPTION
**Pull request by omeshp**

### CODE REVIEW

File: Tasks/DownloadExternalBuildArtifacts/download.ts [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+            let downloader = new engine.FetchEngine();
+            let downloaderOptions = new engine.FetchEngineOptions();
+            await downloader.fetchItems(webProvider, downloadPath, downloaderOptions);
+        }
+        else if (artifact.resource.type.toLowerCase() === "filepath") {
+            // todo: call into robocopy
+            throw new Error("Not Supported");
+        }
+        else {
+            throw new Error("Not Supported");
```
Comment (GitHubSriramB) [inline Tasks/DownloadExternalBuildArtifacts/download.ts]: I believe we used to warn and not throw error if artifact type is not supported. pl. check.

File: Tasks/DownloadPackage/task.json [hunk 2]
```diff
@@ -9,7 +9,7 @@
   "version": {
     "Major": 0,
     "Minor": 1,
-    "Patch": 7
+    "Patch": 0
```
Comment (GitHubSriramB) [inline Tasks/DownloadPackage/task.json]: why is this change required?

File: Tasks/DownloadExternalBuildArtifacts/download.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        }
+    });
+
+    tl.setResult(tl.TaskResult.Succeeded, "");
```
Comment (GitHubSriramB) [inline Tasks/DownloadExternalBuildArtifacts/download.ts]: we should add task telemetry to indicate failures, differentiate user from system and track system errors as part of reliability

File: Tasks/DownloadExternalBuildArtifacts/download.ts [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+            }
+
+            let containerId: number = parseInt(containerParts[1]);
+            let containerPath: string = containerParts[2];
+
+            var itemsUrl = url.resolve(endpointUrl, "_apis/resources/Containers/" + containerId + "?itemPath=" + containerPath + "&isShallow=true");
+            console.log(tl.loc("DownloadArtifacts", itemsUrl));
+
+            var variables = {};
+            var webProvider = new providers.WebProvider(itemsUrl, "vsts.handlebars", "", accessToken, variables);
```
Comment (GitHubSriramB) [inline Tasks/DownloadExternalBuildArtifacts/download.ts]: shouldn't handlebar be part of task rather than being part of ILD library?

File: Tasks/DownloadExternalBuildArtifacts/download.ts [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+            let downloader = new engine.FetchEngine();
+            let downloaderOptions = new engine.FetchEngineOptions();
```
Comment (GitHubSriramB) [inline Tasks/DownloadExternalBuildArtifacts/download.ts]: I believe there are some variables to override parallel limit. we should use that if set.

**Comment by bryanmacfarlane** [comment]
Should this be named "External" since you're downloading from VSTS build?  From RMs perspective it's external but from the customer's perspective this is all VSTS.  I would think external would be things like artifactory etc...

**Comment by GitHubSriramB** [comment]
@bryanmacfarlane - We will bundle this task with External TFS Artifact extension we have. It will not be added to task catalog. This is essentially to download artifacts (Build/Git repo etc.) from a TFS instance or VSTS account different from the one used to Release.

File: Tasks/DownloadExternalBuildArtifacts/task.json [hunk 6]
```diff
@@ -0,0 +1,112 @@
+{
```
Comment (GitHubSriramB) [inline Tasks/DownloadExternalBuildArtifacts/task.json]: We should add this task into vsts-rm-extensions repo & make sure it gets bundled with External TFS Artifacts extension.

File: Tasks/DownloadJenkinsArtifacts/task.json [hunk 7]
```diff
@@ -0,0 +1,91 @@
+{
```
Comment (GitHubSriramB) [inline Tasks/DownloadJenkinsArtifacts/task.json]: Per our discussion,, we should enhance existing "Download Jenkins artifact" task instead of adding another task.

File: Tasks/DownloadTeamCityArtifacts/task.json [hunk 8]
```diff
@@ -0,0 +1,112 @@
+{
```
Comment (GitHubSriramB) [inline Tasks/DownloadTeamCityArtifacts/task.json]: This task should also live in vsts-rm-extensions & get bundled with "TeamCity artifacts" extension

File: Tasks/JenkinsDownloadArtifacts/Strings/resources.resjson/en-US/resources.resjson [hunk 9]
```diff
@@ -14,12 +14,13 @@
   "loc.input.help.jenkinsBuild": "Download artifacts produced by the last successful build, or from a specific build instance.",
   "loc.input.label.jenkinsBuildNumber": "Jenkins build number",
   "loc.input.help.jenkinsBuildNumber": "Download artifacts produced by this build.",
+  "loc.input.label.itemPattern": "Item Pattern",
+  "loc.input.help.itemPattern": "Minmatch pattern to filter files to be downloaded.",
```
Comment (GitHubSriramB) [inline Tasks/JenkinsDownloadArtifacts/Strings/resources.resjson/en-US/resources.resjson]: typo: Minimatch

File: Tasks/JenkinsDownloadArtifacts/jenkinsdownloadartifacts.ts [hunk 10]
```diff
@@ -134,18 +96,30 @@ async function doWork() {
...
+        var variables = {
+            "endpoint": {
+                "url": serverEndpointUrl
+            },
+            "definition": jobName,
+            "version": buildId
+        };
+        var handler = new handlers.BasicCredentialHandler(username, password);
+        
+        var webProvider = new providers.WebProvider(artifactQueryUrl, templatePath, variables, handler, { ignoreSslError: strictSSL });
```
Comment (GitHubSriramB) [inline Tasks/JenkinsDownloadArtifacts/jenkinsdownloadartifacts.ts]: ignoreSslError getting assigned to strictSSL seems incorrect. should we instead have the parameter also as strictSSL?

File: Tasks/JenkinsDownloadArtifacts/task.json [hunk 11]
```diff
@@ -68,15 +71,44 @@
...
+            }
+        },
+        {
+          "name": "itemPattern",
+          "type": "string",
+          "label": "Item Pattern",
+          "defaultValue": "**",
+          "required": false,
+          "groupName": "advanced",
+          "helpMarkDown": "Minmatch pattern to filter files to be downloaded."
```
Comment (GitHubSriramB) [inline Tasks/JenkinsDownloadArtifacts/task.json]: typo - Minimatch

File: Tasks/JenkinsDownloadArtifacts/jenkinsdownloadartifacts.ts [hunk 12]
```diff
@@ -134,18 +96,30 @@ async function doWork() {
...
+        var localFileProvider = new providers.FilesystemProvider(localPathRoot);
+    
+        let downloader = new engine.ArtifactEngine();
+        var downloaderOptions = new engine.ArtifactEngineOptions();
+        downloaderOptions.itemPattern = tl.getInput('itemPattern', false) || "**";
+        downloaderOptions.parallelProcessingLimit = +tl.getVariable("release.artifact.download.parallellimit") || 4;
```
Comment (GitHubSriramB) [inline Tasks/JenkinsDownloadArtifacts/jenkinsdownloadartifacts.ts]: the + at the start is to cast to integer?

File: Tasks/JenkinsDownloadArtifacts/jenkins.handlebars.txt [hunk 13]
```diff
@@ -0,0 +1,11 @@
+[
+{{#artifacts}}
+  {
+    "path" : "{{this.relativePath}}",
+    "itemType": "file",
+    "metadata" : {
+      "downloadUrl": "{{../endpoint.url}}/job/{{../definition}}/{{../version}}/artifact/{{this.relativePath}}"
```
Comment (GitHubSriramB) [inline Tasks/JenkinsDownloadArtifacts/jenkins.handlebars.txt]: what do the ../ in the url signify?

Comment (omeshp) [reply Tasks/JenkinsDownloadArtifacts/jenkins.handlebars.txt]: currently we are in artifacts context, we need to go up one level where variables are available. Just navigating the jobject.

File: Tasks/JenkinsDownloadArtifacts/jenkinsdownloadartifacts.ts (hunk 10 above)
Comment (omeshp) [reply Tasks/JenkinsDownloadArtifacts/jenkinsdownloadartifacts.ts]: yes it should be inverted, ignoreSslError is defined by typed-rest-client.

File: Tasks/JenkinsDownloadArtifacts/jenkinsdownloadartifacts.ts (hunk 12 above)
Comment (omeshp) [reply Tasks/JenkinsDownloadArtifacts/jenkinsdownloadartifacts.ts]: yes

File: Tasks/JenkinsDownloadArtifacts/task.json [hunk 14]
```diff
@@ -39,11 +39,14 @@
         },
         {
             "name": "jobName",
-            "type": "string",
+            "type": "pickList",
```
Comment (madhurig) [inline Tasks/JenkinsDownloadArtifacts/task.json]: Isn't updating string to pickList a breaking change? What if the user has something that can't be retrieved by the data bindings?

Comment (omeshp) [reply Tasks/JenkinsDownloadArtifacts/task.json]: It works fine since the pickList is editable, user can provide any other value as input also.
